### PR TITLE
feat(fulltext_index): allow enable full-text index in SQL and gRPC way

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#4246 

depends on https://github.com/GreptimeTeam/greptime-proto/pull/179 

## What's changed and what's your intention?

* SQL Parser supports `FULLTEXT` syntax
* Pass full-text options to gRPC requests
* Set full-text options for pipeline

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced the handling of fulltext options in column schemas with the addition of new methods and structs to manage these options.
  - Enhanced the `CreateTable` and `CreateExternalTable` structures with a new `Column` struct to encapsulate column properties.

- **Bug Fixes**
  - Standardized error handling for column data types and fulltext index validation.

- **Documentation**
  - Updated error variants and function signatures to reflect new fulltext options and column properties.

- **Refactor**
  - Renamed keys and added new methods to handle fulltext options in `ColumnSchema`.
  - Refactored column-related structures and function signatures to improve readability and maintainability.

### CODE REVIEW

File: src/operator/src/expr_factory.rs [hunk 1]
```diff
@@ -34,13 +34,13 @@ use query::sql::{
 use session::context::QueryContextRef;
 use session::table_name::table_idents_to_full_name;
 use snafu::{ensure, OptionExt, ResultExt};
-use sql::ast::{ColumnDef, ColumnOption, TableConstraint};
+use sql::ast::{ColumnOption, TableConstraint};
 use sql::statements::alter::{AlterTable, AlterTableOperation};
 use sql::statements::create::{
-    CreateExternalTable, CreateFlow, CreateTable, CreateView, TIME_INDEX,
+    Column as SqlColumn, CreateExternalTable, CreateFlow, CreateTable, CreateView, TIME_INDEX,
```
[the remark this answers is not in the document]
Comment (killme2008) [reply src/operator/src/expr_factory.rs:40]: @coderabbitai This consistency is unnecessary, please don't suggest it again.

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/sql/src/error.rs [hunk 2]
```diff
@@ -252,6 +252,20 @@ pub enum Error {
         #[snafu(implicit)]
         location: Location,
     },
+
+    #[snafu(display("Invalid fulltext option: {}", msg))]
+    FulltextInvalidValue {
```
Comment (waynexia) [inline src/sql/src/error.rs:257]: Is this `FulltextInvalidOption`?

File: src/api/src/v1/column_def.rs [hunk 3]
```diff
@@ -43,13 +48,43 @@ pub fn try_as_column_schema(column_def: &ColumnDef) -> Result<ColumnSchema> {
...
+
+    ColumnSchema::new(&column_def.name, data_type.into(), column_def.is_nullable)
+        .with_metadata(metadata)
+        .with_time_index(column_def.semantic_type() == SemanticType::Timestamp)
+        .with_default_constraint(constraint)
+        .context(error::InvalidColumnDefaultConstraintSnafu {
+            column: &column_def.name,
+        })
+}
+
```
Comment (killme2008) [inline src/api/src/v1/column_def.rs:65]: Could you add docs for these `pub` functions?

Comment (killme2008) [reply src/api/src/v1/column_def.rs:65]: Adds some unit tests if possible.

File: src/common/grpc-expr/src/error.rs [hunk 4]
```diff
@@ -104,6 +105,25 @@ pub enum Error {
...
+    UnknownColumnDataType {
+        datatype: i32,
+        #[snafu(implicit)]
+        location: Location,
+        #[snafu(source)]
+        error: prost::DecodeError,
+    },
+
+    #[snafu(display(
+        "Full-text index only supports string type, column: {column_name}, unexpected type: {column_type:?}"
```
Comment (killme2008) [inline src/common/grpc-expr/src/error.rs:119]: To maintain consistency.

```suggestion
        "Fulltext index only supports string type, column: {column_name}, unexpected type: {column_type:?}"
```

File: src/sql/src/statements/create.rs [hunk 5]
```diff
@@ -83,14 +83,62 @@ pub struct CreateTable {
...
-    pub columns: Vec<ColumnDef>,
+    pub columns: Vec<Column>,
     pub engine: String,
     pub constraints: Vec<TableConstraint>,
     /// Table options in `WITH`. All keys are lowercase.
     pub options: OptionMap,
     pub partitions: Option<Partitions>,
 }
 
+#[derive(Debug, PartialEq, Eq, Clone, Visit, VisitMut)]
```
Comment (killme2008) [inline src/sql/src/statements/create.rs:94]: Please add docs

File: src/sql/src/statements.rs [hunk 6]
```diff
@@ -389,6 +394,40 @@ pub fn column_def_to_schema(
             .insert(COMMENT_KEY.to_string(), c.to_string());
     }
 
+    if let Some(fulltext_opts) = &column.extensions.fulltext_options {
```
Comment (killme2008) [inline src/sql/src/statements.rs:397]: What about extracting this block into a function like `with_fulltext_options`?  Make this function smaller.

**Review by killme2008 — APPROVED** [review]
LGTM. Thank you.

**Pull request merged**